pmlc/ast/builder.cc

```diff
@@ -738,6 +739,38 @@ struct ProgramBuilder {
         .result();
   }
 
+  Value makeGatherOp(ExprNodeIntrinsic *node, ArrayRef<Value> operands) {
+    TensorShape shape = evaluator.getShape(node);
+    RankedTensorType resultType = builder.getRankedTensorType(shape);
+    auto op = builder.create<tile::GatherOp>(
+        loc, resultType, ValueRange{operands[0], operands[1]},
+        ArrayRef<NamedAttribute>{});
+
+    FloatAttr cubicCoeff;
+    if (!m_Constant(&cubicCoeff).match(operands[4].getDefiningOp())) {
+      throw std::runtime_error(
+          "'cubicCoeff' primitive expects argument 5 to be a constant float");
+    }
+    auto cubicCoeffFloat = static_cast<float>(cubicCoeff.getValueAsDouble());
+    op.setAttr("cubicCoeff", builder.getF32FloatAttr(cubicCoeffFloat));
+
+    IntegerAttr mode;
+    if (!m_Constant(&mode).match(operands[3].getDefiningOp())) {
+      throw std::runtime_error(
+          "'mode' primitive expects argument 4 to be a constant integer");
+    }
+    op.setAttr("mode", mode);
+
+    IntegerAttr axis;
+    if (!m_Constant(&axis).match(operands[2].getDefiningOp())) {
+      throw std::runtime_error(
+          "'axis' primitive expects argument 3 to be a constant integer");
```

Comment:
All of these validation checks should happen earlier, so that if an exception is thrown, the call stack will eventually trace back to the line where the eDSL op was constructed. By throwing it here, the call stack will look like the error happened in `buildProgram`, which will be harder for users to determine where the actual error in their own code was.

To fix this, do as much validation as possible in the `ast_ops.cc` code.

Reply:
Also, the error message here should be more like:

"'gather' primitive expects the 'axis' argument to be a constant integer"

Reply:
Fixed.